### PR TITLE
Scripts -> makefile 

### PR DESCRIPTION
**Pull request by Hedingber**
Mainly inspired by https://github.com/nuclio/nuclio/blob/development/Makefile

### CODE REVIEW

File: dockerfiles/models-gpu/Dockerfile [hunk 1]
```diff
@@ -149,7 +149,7 @@ RUN python -c "from matplotlib.font_manager import _rebuild; _rebuild();"
 
 RUN conda clean -aqy
 
-RUN python -m pip install git+https://github.com/${REPO}/mlrun.git@${MLRUN_TAG}
+RUN python -m pip install git+https://github.com/${MLRUN_GITHUB_REPO}/mlrun.git@${MLRUN_PACKAGE_TAG}
 
 RUN python -m pip install --ignore-installed --no-cache-dir \
```
Comment (liranbg) [inline dockerfiles/models-gpu/Dockerfile:154]: looks like same requirement as `base`, have a common requirement.txt to unite them all and install when needed

File: dockerfiles/models/Dockerfile [hunk 2]
```diff
@@ -119,7 +119,7 @@ RUN cd /repos && \
 
 RUN python -c "from matplotlib.font_manager import _rebuild; _rebuild();"
 
-RUN python -m pip install git+https://github.com/${REPO}/mlrun.git@${MLRUN_TAG}
+RUN python -m pip install git+https://github.com/${MLRUN_GITHUB_REPO}/mlrun.git@${MLRUN_PACKAGE_TAG}
 
 
 # remove dask verify version mismatch warnings
```
Comment (liranbg) [inline dockerfiles/models/Dockerfile:125]: again here, versions are all over -> recommend moving under a single requirement.txt file

File: Makefile [hunk 3]
```diff
@@ -12,18 +12,179 @@
...
+
+.PHONY: docker-images
+docker-images: $(DOCKER_IMAGES_RULES)
+	@echo Done.
+
+
+.PHONY: push-docker-images
+push-docker-images: docker-images
+	for image in $(IMAGES_TO_PUSH); do \
+		docker push $$image ; \
```
[the remark this answers is not in the document]
Comment (liranbg) [reply Makefile:56]: also, please consider parallelizing this command as it is IO bounded
you may do it with `xargs`
e.g.:
`@echo $IMAGES_TO_PUSH | xargs -n 1 docker push`

File: Makefile [hunk 4]
```diff
@@ -12,18 +12,179 @@
...
 
+MLRUN_DOCKER_TAG := $(if $(MLRUN_DOCKER_TAG),$(MLRUN_DOCKER_TAG),latest)
+MLRUN_DOCKER_REPO := $(if $(MLRUN_DOCKER_REPO),$(MLRUN_DOCKER_REPO),mlrun)
+MLRUN_DOCKER_REGISTRY := $(if $(MLRUN_DOCKER_REGISTRY),$(MLRUN_DOCKER_REGISTRY),) #e.g. quay.io/
+MLRUN_ML_DOCKER_IMAGE_NAME_PREFIX := $(if $(MLRUN_ML_DOCKER_IMAGE_NAME_PREFIX),$(MLRUN_ML_DOCKER_IMAGE_NAME_PREFIX),ml-)
+MLRUN_PACKAGE_TAG := $(if $(MLRUN_PACKAGE_TAG),$(MLRUN_PACKAGE_TAG),development)
+MLRUN_GITHUB_REPO := $(if $(MLRUN_GITHUB_REPO),$(MLRUN_GITHUB_REPO),mlrun)
+MLRUN_ML_PYTHON_VERSION := $(if $(MLRUN_ML_PYTHON_VERSION),$(MLRUN_ML_PYTHON_VERSION),3.8)
+MLRUN_LEGACY_ML_PYTHON_VERSION := $(if $(MLRUN_LEGACY_ML_PYTHON_VERSION),$(MLRUN_LEGACY_ML_PYTHON_VERSION),3.6)
+MLRUN_API_PYTHON_VERSION := $(if $(MLRUN_API_PYTHON_VERSION),$(MLRUN_API_PYTHON_VERSION),3.7)
```
Comment (liranbg) [inline Makefile:23]: perhaps use the same version across ML & API?

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hedingber) [reply Makefile:56]: the `;` is needed, note the `\` at the end of every line, this whole for loop is actually executed as one liner - so the `;` is needed

[the remark this answers is not in the document]
Comment (Hedingber) [reply Makefile:56]: I would prefer to wait with parallelizing this for now
BTW running it as you suggested will not achieve parallelization, you need to set the `-P` to something greater than 1

File: Makefile [hunk 5]
```diff
@@ -12,18 +12,179 @@
...
+	for image in $(IMAGES_TO_PUSH); do \
+		docker push $$image ; \
+	done
+	@echo Done.
+
+
+.PHONY: print-docker-images
+print-docker-images:
+	@for image in $(IMAGES_TO_PUSH); do \
+		echo $$image ; \
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply Makefile:64]: needed, see above comment

File: Makefile [hunk 6]
```diff
@@ -32,41 +193,41 @@ clean:
 	find . -name '*.pyc' -exec rm {} \;
 
 
+.PHONY: test-dockerized
+test-dockerized: build-test
+	-docker network create mlrun
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply Makefile:199]: I cannot, just copy-pasted, there is a tests refactor coming which will probably remove this if it's really redundant

File: dockerfiles/README.md [hunk 7]
```diff
@@ -1,32 +1,33 @@
...
 for example,
-  `./docker-build.sh mlrun ml v0.4.7 0.4.6 3.7 3.6`
+  `MLRUN_DOCKER_TAG=0.4.6 MLRUN_DOCKER_REPO=mlrun MLRUN_ML_DOCKER_IMAGE_NAME_PREFIX=ml- MLRUN_PACKAGE_TAG=v0.4.7 MLRUN_GITHUB_REPO=mlrun MLRUN_ML_PYTHON_VERSION=3.7 MLRUN_API_PYTHON_VERSION=3.6 make docker-images`
 this will generate the following images:
-  * `mlrun/ml-base:0.4.7`       (python 3.7)
-  * `mlrun/ml-models:0.4.7`     (python 3.7)
-  * `mlrun/ml-models-gpu:0.4.7` (python 3.7) 
-  * `mlrun/mlrun-api:0.4.7`     (python 3.7)
-  * `mlrun/mlrun:0.4.7`         (python 3.7)
+  * `mlrun/ml-base:0.4.6`       (python 3.7)
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply dockerfiles/README.md:20]: Nope, `MLRUN_DOCKER_TAG` is 0.4.6

File: dockerfiles/base/Dockerfile [hunk 8]
```diff
@@ -28,7 +28,7 @@ RUN apt update -qqq \
 
 ENV SSL_CERT_DIR /etc/ssl/certs
 
-RUN python -m pip install git+https://github.com/${REPO}/mlrun.git@${MLRUN_TAG}
+RUN python -m pip install git+https://github.com/${MLRUN_GITHUB_REPO}/mlrun.git@${MLRUN_PACKAGE_TAG}
 
 RUN python -m pip install urllib3==1.25.8  kubernetes==10.0.1
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply dockerfiles/base/Dockerfile:33]: True, @yjb-ds has WIP on these docker files, I passed him your suggestion

File: dockerfiles/models-gpu/Dockerfile (hunk 1 above)
Comment (Hedingber) [reply dockerfiles/models-gpu/Dockerfile:154]: @yjb-ds this is for you
@liranbg see above comment about requirements,

File: dockerfiles/models/Dockerfile (hunk 2 above)
Comment (Hedingber) [reply dockerfiles/models/Dockerfile:125]: same @yjb-ds

File: Makefile (hunk 4 above)
Comment (Hedingber) [reply Makefile:23]: Needed @yjb-ds approval for this - approved, we will have:
`MLRUN_PYTHON_VERSION` and `MLRUN_LEGACY_ML_PYTHON_VERSION`

**Pull request merged**